File: bench/apertures/apertures.jl
using Photometry
using BenchmarkTools
using CSV
using DataFrames
using ProgressLogging

AP = [
    CircularAperture,
    CircularAnnulus,
    EllipticalAperture,
    EllipticalAnnulus,
    RectangularAperture,
    RectangularAnnulus,
]
PARAMS = [
    (3),
    (3, 5),
    (3, 3, 0),
    (3, 5, 4, 0),
    (3, 5, 0),
    (3, 5, 4, 0)
]

data = randn(512, 512) .+ 10

rows = []

@progress for N in [1, 10, 50, 100], method in [:exact, :center, (:subpixel, 5)], (A, par) in zip(AP, PARAMS)
    aper = A(256, 256, par...)
    apers = repeat([aper], N)
    func(a, d) = aperture_photometry(a, d, method=method)
    time = @belapsed $func($apers, $data)
    m = method isa Tuple ? "$(String(method[1]))-$(method[2])" : String(method)
    push!(rows, (N=N, aperture=A, method=m, time=time))
end

results = DataFrame(rows)
CSV.write(joinpath(@__DIR__, "julia_apertures.csv"), results)

File: benchmarks/apertures.py
from photutils import (
    CircularAperture,
    CircularAnnulus,
    EllipticalAperture,
    EllipticalAnnulus,
    RectangularAperture,
    RectangularAnnulus,
    aperture_photometry,
)
from datetime import datetime
import pandas as pd
import numpy as np
from tqdm import tqdm
from itertools import product, repeat
import os

APS = [
    CircularAperture,
    CircularAnnulus,
    EllipticalAperture,
    EllipticalAnnulus,
    RectangularAperture,
    RectangularAnnulus,
]
params = [(3,), (3, 5), (3, 3, 0), (3, 5, 4, 0), (3, 5, 0), (3, 5, 4, 0)]

Ns = [1, 10, 50, 100]
methods = ["exact", "center", "subpixel"]

data = np.random.randn(512, 512) + 10

inputs = list(product(Ns, methods, zip(APS, params)))

rows = []
for input_ in tqdm(inputs):
    N, method, (A, par) = input_
    if method == "exact" and (A == RectangularAnnulus or A == RectangularAperture):
        continue
    aps = A(list(repeat((255, 255), N)), *par)
    m = method if method != "subpixel" else "subpixel-5"
    kwargs = {} if method != "subpixel" else {"subpixels": 5}
    t0 = datetime.now()
    t = aperture_photometry(data, aps, method=method, **kwargs)
    t1 = datetime.now()
    time = (t1 - t0).total_seconds()
    rows.append((N, A.__name__, m, time))

df = pd.DataFrame(rows, columns=["N", "aperture", "method", "time"])

path = os.path.dirname(__file__)
df.to_csv(os.path.join(path, "python_apertures.csv"), index=False)
